### ISSUE TITLE
Legion: Final Tracker

### ISSUE THREAD

**Issue by AlveLarsson** [amethyst/amethyst#2517]
Hello! I've been out of the loop for the past months due to some life issues, but now I'm back.

Progress on the Legion port has slowed to a standstill.  I'm creating this issue to track what's been done, what still needs to be done, and what our roadblocks are. This issue will act as a new, clean slate to minimize confusion and get us back on track.

_Edit this comment to change the lists below, or comment below and someone with write access will update the lists for you._

#### Current Progress with Names & Links
- [Project Board](https://github.com/amethyst/amethyst/projects/21)
- [Release Checklist](https://github.com/amethyst/amethyst/issues/2308)
- [x] amethyst parent crate #2320
- [ ] examples #2483
- [ ] book
- [ ] inline documentation (please do it _with_ your subcrate ports!)
- [ ] all subcrates #2400 #2417 #2436
  - [x] amethyst_animation #2334 #2351
  - [x] amethyst_assets @jlowry #2329 #2404 #2520
  - [x] amethyst_audio #2315
  - [x] amethyst_config
  - [x] amethyst_controls #2325 #2378
  - [x] amethyst_core @LechintanTudor #2502 #2560
  - [x] amethyst_derive
  - [x] amethyst_error
  - [x] amethyst_gltf #2335 #2631
  - [x] amethyst_input #2332 #2481
  - [x] amethyst_locale #2331 #2336
  - [x] amethyst_network #2319
  - [x] amethyst_rendy #2318 #2522 
  - [ ] amethyst_test
  - [x] amethyst_tiles @valkum #2555
  - [x] amethyst_ui @grzi  #2330 #2556 
  - [x] amethyst_utils #2350
  - [x] amethyst_window

#### Who is willing to help? What are you currently working on?
- @AlveLarsson - tracking, talking to folks to gather updates
- @CleanCut - help with tracking, try to make `amethyst_rendy` less of a pain, maybe `examples`
- @jlowry - `amethyst_assets`
- @valkum - `amethyst_tiles`
- @LechintanTudor - `amethyst_core`, `amethyst_ui`

Lastly, we need to communicate frequent updates about what's being done or rather the rate of progress that's happening on each crate as we start work.

**Important!** Please discuss the legion port on GitHub issues (like this one, or more specific ones), not on discord or the forum. We need to keep the discussion in one place to keep things sane.

(Special thanks to @CleanCut and @jlowry for catch-ups)

**Comment by valkum** [comment]
In the meantime legion progressed to 0.3.1.
How much time do we need to spent to move the currently finished migrations to this version?
Should we consider this as an option?

**Comment by jlowry** [comment]
I believe that the work for UI #2330  may already have been done by @LechintanTudor according to this [discord comment](https://discord.com/channels/425678876929163284/425679992244928512/758736008035696710) in [https://github.com/LechintanTudor/amethyst_ui_legion/tree/legion_v2](https://github.com/LechintanTudor/amethyst_ui_legion/tree/legion_v2).

**Comment by jlowry** [comment]
I am currently working working on #2404 Prefab system in legion.
I've been working on integrating [Atelier Assets](https://github.com/amethyst/atelier-assets) and [Legion Prefab](https://github.com/kabergstrom/prefab/).
On [my fork's `assets_wip` branch](https://github.com/jlowry/amethyst/tree/assets_wip), I have atelier integrated and legion prefab working for non nested prefabs (also the system that applys the prefabs automatcally is not done.). This is still needs tidying up but I'm making steady progress.

Currently, I am working on format part of legion prefab based recommendations by @kabergstrom. To improve the serialisation format to make easier to use when writing prefabs by hand, and so that atelier asset handles can be created automatically for nested prefabs and assets within them.

**Comment by jlowry** [comment]
I believe this [card 'Reimplement pausable systems (system_ext.rs)'](https://github.com/amethyst/amethyst/projects/21#card-43449365) is done in this pull request **Migrate SystemExt pauseable to Legion #2467** by @thunderseethe.

**Comment by CleanCut** [comment]
**Everyone:** For work in progress, please branch off of `legion_v2` (or something downstream from it) and make a _Draft Pull Request_ for your branch/fork against `legion_v2` while you do your work. That way we have a PR to link to, see progress as it occurs, and communicate with you on.  Switch the PR out of draft mode when you're ready for a review.

- @BlockCat I see you were working on `amethyst_animation` in https://github.com/amethyst/amethyst/pull/2351 - would you be willing to pick that back up and merge `legion_v2` into that branch? Should we add you to the active list?
- @chemicstry I don't see anything "in progress" from you, but you've been very involved in the past. Should we add you to the active list? Do you have something in progress I'm unaware of? 
- @jlowry As per your comment above I've listed you as the person for `amethyst_assets` and linked to your PR.
- @jojolepro In Discord you said you were not participating any further. If you change your mind, we'd love to have you back! If you are willing to give a summary of where you left off, that would be helpful.
- @LechintanTudor Do you have a PR out for `amethyst_ui` somewhere to supersede https://github.com/amethyst/amethyst/pull/2343? Should I add you to the active list? You mentioned you were working on `amethyst_core` in Discord as well - but I don't see a PR for that, either - are you working on both in master in your fork? Should we add you to the active list?
- @mockersf Would you like to finish up `examples`? At the point https://github.com/amethyst/amethyst/pull/2483 was left off, it looks like were CI failures to address. Should I add you to the active list?
- @thunderseethe I see you had been working on part of `amethyst_core` in https://github.com/amethyst/amethyst/pull/2467 - shall we add you to the active list to finish that up?

I'm also going to shotgun a question to a bunch of notable folks I'm aware of who have been involved in some way in the past -> Would any of you folks like to directly contribute to this effort to finish up and ship Legion integration in Amethyst?  Feel free to ignore this ping and unsubscribe from this issue if you're not interested.
- @ab0v3g4me @aclysma @azriel91 @Blisto91 @erlend-sh @fhaynes @kabergstrom (atelier-assets) @jaynus @Moxinilian @Rhuagh @kabergstrom @termhn @TomGillen @Xaeroxe 

I've surely missed some people from these lists. (Sorry--unintentional!) Please pass this along to them! Please speak up if you're willing to contribute.

**Comment by CleanCut** [comment]
> I believe this [card 'Reimplement pausable systems (system_ext.rs)'](https://github.com/amethyst/amethyst/projects/21#card-43449365) is done in this pull request **Migrate SystemExt pauseable to Legion #2467** by @thunderseethe.

👍 Archived the card in favor of the PR.

**Comment by LechintanTudor** [comment]
@CleanCut I'm willing to port amethyst_core to the latest version of legion. I'll set up a draft pull request tomorrow.

Regarding amethyst_ui, the port to legion 0.2 was mostly finished but I prefer to update it to 0.3 after the other crates are finished.

**Comment by CleanCut** [comment]
@valkum I missed you in my big comment since you were already on the active list. What are you working on?

**Comment by CleanCut** [comment]
@valkum @jlowry I'm not able to compile the current `legion_v2` branch on my mac.  Most (but not all) of the errors are very similar to this:

```
error[E0405]: cannot find trait `SystemDesc` in crate `amethyst_core`
   --> amethyst_derive/tests/system_desc.rs:140:21
    |
140 |     #[derive(Debug, SystemDesc)]
    |                     ^^^^^^^^^^ not found in `amethyst_core`
    |
    = note: this error originates in a derive macro (in Nightly builds, run with -Z macro-backtrace for more info)
```

**Comment by CleanCut** [comment]
I'm understanding more now, but not enough to get unblocked.

It looks like the way the legion project was approached was to replace `specs` with `legion` in `amethyst_core` all at once, breaking the world, and now subcrates are being fixed one by one.  So CI and workspace-wide builds are both non-functional 🙁 .

I figure the way to see if individual crates are working is to run things like
- `cargo build [--features=x]`
- `cargo test [--features=x]`

But crates like `amethyst_utils` and `amethyst_audio` don't support backend graphics "features", yet at the same time weirdly depend on `amethyst_rendy`...which requires selecting a graphical backend feature.  But I can't find a way to pass a graphical backend feature when compiling something like `amethyst_utils` alone. What am I missing?

```shell
amethyst_utils $ cargo build
   Compiling amethyst_rendy v0.15.3 (/Users/nathan/am/amethyst/amethyst_rendy)
error: You must specify at least one graphical backend feature: "metal", "vulkan", "empty"
"See the wiki article https://book.amethyst.rs/stable/appendices/c_feature_gates.html#graphics-features for more details."
   --> amethyst_rendy/src/types.rs:111:1
    |
111 | / impl_backends!(
112 | |     // DirectX 12 is currently disabled because of incomplete gfx-hal support for it.
113 | |     // It will be re-enabled when it actually works.
114 | |     // Dx12, "dx12", rendy::dx12::Backend;
...   |
117 | |     Empty, "empty", rendy::empty::Backend;
118 | | );
    | |__^
    |
    = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to previous error

error: could not compile `amethyst_rendy`.

To learn more, run the command again with --verbose.

amethyst_utils $ cargo build --features=metal
error: Package `amethyst_utils v0.15.3 (/Users/nathan/am/amethyst/amethyst_utils)` does not have these features: `metal`
```

The only way I can think of to resolve this is to add `metal`, `vulkan`, etc. features to everything, which seems like a terrible solution. Is that what people have been doing?

**Comment by jlowry** [comment]
On branch legion_v2 [22e0f3bc4](https://github.com/amethyst/amethyst/commit/22e0f3bc4cfa309b6d279e49969733d76ba1c6b5) is HEAD.

For building the amethyst root crate I run:
`cargo build --features "metal"`

To test things are working, I've been running an example, e.g.:

`cargo run --features "metal" --example "asset_custom"`

It doesn't solve your problem of crates not compiling with unrelated featurs. But I find it useful to see what is compiling from the dependency graph (`cargo tree | grep amethyst`) :
`
amethyst v0.15.3
├── amethyst_assets v0.15.3
│   ├── amethyst_core v0.15.3
│   │   ├── amethyst_error v0.15.3
│   ├── amethyst_derive v0.15.3
│   ├── amethyst_error v0.15.3
├── amethyst_audio v0.15.3
│   ├── amethyst_assets v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_error v0.15.3
├── amethyst_config v0.15.3
├── amethyst_controls v0.15.3
│   ├── amethyst_assets v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_derive v0.15.3
│   ├── amethyst_error v0.15.3
│   ├── amethyst_input v0.15.3
│   │   ├── amethyst_config v0.15.3
│   │   ├── amethyst_core v0.15.3
│   │   ├── amethyst_error v0.15.3
│   │   ├── amethyst_window v0.15.3
│   │   │   ├── amethyst_config v0.15.3
│   │   │   ├── amethyst_core v0.15.3
│   │   │   ├── amethyst_error v0.15.3
├── amethyst_core v0.15.3
├── amethyst_derive v0.15.3
├── amethyst_error v0.15.3
├── amethyst_input v0.15.3
├── amethyst_locale v0.15.3
│   ├── amethyst_assets v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_error v0.15.3
├── amethyst_network v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_error v0.15.3
├── amethyst_rendy v0.15.3
│   ├── amethyst_assets v0.15.3
│   ├── amethyst_config v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_derive v0.15.3
│   ├── amethyst_error v0.15.3
│   ├── amethyst_window v0.15.3
├── amethyst_utils v0.15.3
│   ├── amethyst_assets v0.15.3
│   ├── amethyst_core v0.15.3
│   ├── amethyst_derive v0.15.3
│   ├── amethyst_error v0.15.3
│   ├── amethyst_rendy v0.15.3
│   ├── amethyst_window v0.15.3
├── amethyst_window v0.15.3
`

**Comment by Blisto91** [comment]
Sorry for dropping off the the face of the earth.

Due to some health issues I haven't had the energy or motivation to be active.
And it's not realistic that I'll be able to help soon even tho I want to. So the monthly donation will have to do for now :)
Still gonna lurk in the background and follow progress when I can.

**Comment by jlowry** [comment]
My draft pull request is #2520 but I do not have permission to link it against my issue #2404.

**Comment by CleanCut** [comment]
> Sorry for dropping off the the face of the earth.
> 
> Due to some health issues I haven't had the energy or motivation to be active.
> And it's not realistic that I'll be able to help soon even tho I want to. So the monthly donation will have to do for now :)
> Still gonna lurk in the background and follow progress when I can.

I'm so sorry to hear that! Please take care of yourself and get feeling better!

**Comment by chemicstry** [comment]
I don't have anything in progress and unfortunately I probably won't find time to contribute as I'm very busy with university stuff.

However, feel free to ping me if you need some help or have any issues with my commits

**Comment by LechintanTudor** [comment]
Here's my draft for amethyst_core #2521

**Comment by CleanCut** [comment]
Last night I started working on auto-selecting `vulkan` or `metal` backends based on platform, but I've run into some roadblocks trying to figure out how to get the `empty` feature to override the auto-detected choice in the macros.  I just committed and pushed what I had as a draft PR in https://github.com/amethyst/amethyst/pull/2522. I could use some help figuring this out.

I'll dig into it again this evening.

**Comment by CleanCut** [comment]
> For building the amethyst root crate I run:
> `cargo build --features "metal"`
> 
> To test things are working, I've been running an example, e.g.:
> 
> `cargo run --features "metal" --example "asset_custom"`
> 
> It doesn't solve your problem of crates not compiling with unrelated features. But I find it useful to see what is compiling from the dependency graph (`cargo tree | grep amethyst`) :

@jlowry Awesome, thank you.  That unblocks for reviewing and helping out with other stuff at least.

I still find it maddening not to be able to compile individual crates, so I'm going to continue on my quest to eliminate mandatory features altogether in #2522

**Comment by tigleym** [comment]
@CleanCut I'd be interested in helping out with updating examples or documentation. But I don't know much of this work is currently blocked by some of the pending work in the other subcrates, so perhaps this should wait towards the end?

**Comment by mockersf** [comment]
I think that, before fixing examples, https://github.com/amethyst/legion/issues/178 should be fixed. This bug means that macros from legion can't be used, which makes systems ... less friendly to write, and probably not what the examples should look like.

I proposed a PR for that but now (after seeing how other projects do it) I don't think it's the best fix (see https://github.com/amethyst/legion/pull/186#issuecomment-723663107 for a potential better idea)

**Comment by CleanCut** [comment]
@tigleym There's no reason you couldn't dig into documentation updates immediately for the book, if you're interested in that.  Any progress you make is progress!

**Comment by AlveLarsson** [comment]
Bi-Weekly updates could be found here now
https://community.amethyst.rs/t/legion-transition-bi-weekly-status-update/1736

**Comment by CleanCut** [comment]
I'm leaving my attempt at automatic graphics selection paused at: https://github.com/amethyst/amethyst/pull/2525#issuecomment-738829623

Would anyone like a collaborator on their part of porting? I'd like to pair with someone on something.

**Comment by jlowry** [comment]
@CleanCut  I'd de delighted to have some help.

As per this [Discord comment](https://discord.com/channels/425678876929163284/425679992244928512/783465972799242260) It would be super useful if serde-diff supported:
```
#[serde(from = "TransformValues", into = "TransformValues")]
and
#[serde_diff(opaque)]
```
Serde diff is used by the prefab system to apply changes to prefabs only on the parts that differ.
For example you could have many light entites that are the same but only differ by transform.

**Comment by CleanCut** [comment]
@jlowry I've got a couple hours tomorrow afternoon -- I'll take a look!

**Comment by ezpuzz** [comment]
I've updated the pong examples to get through example 4 working and 5 ready for amethyst_ui new version at #2531

**Comment by valkum** [comment]
Got tiles nearly ready but got stuck with the camera_switch in the tiles example: https://github.com/valkum/amethyst/blob/legion_v2_tiles/examples/tiles/main.rs#L141 
I am in a state where the example runs but without the functionality implemented in the camera_switch system.

@CleanCut I would appreciate your input if you got time, hit me up on discord too explain further.

**Comment by ezpuzz** [comment]
@valkum thanks for this!  I'm really glad to see an example of using the legion system macros inside an amethyst project.

**Comment by grzi** [comment]
Doing an amethyst_ui port, you can follow my progress in my tracking issue : https://github.com/grzi/amethyst/issues/1

**Comment by CleanCut** [comment]
> As per this [Discord comment](https://discord.com/channels/425678876929163284/425679992244928512/783465972799242260) It would be super useful if [`serde-diff`](https://github.com/amethyst/serde-diff/)supported:
> 
> ```
> #[serde(from = "TransformValues", into = "TransformValues")]
> and
> #[serde_diff(opaque)]
> ```

PR is ready for review in https://github.com/amethyst/serde-diff/pull/22

**Comment by CleanCut** [comment]
Here's a quick status roundup of what I see with a quick look around (I'm sure I've missed some things):

- [Deferred] @cleancut Shelved efforts to eliminate the `--features metal|vulkan` requirement until an upstream feature is stabilized. https://github.com/amethyst/amethyst/pull/2525
- [Needs Review] @cleancut Wrote up feature for `serde_diff` to support target structs to unblock 👇🏼  - https://github.com/amethyst/serde-diff/pull/22
- [In Progress] @jlowry working on the prefab system https://github.com/amethyst/amethyst/pull/2520
- [Merged] @ezpuzz Fixed the `sprites_ordered` example and some docs examples in `amethyst_core` https://github.com/amethyst/amethyst/pull/2535 https://github.com/amethyst/amethyst/pull/2542
- [In Progress] @mockersf @ezpuzz Working on various examples. #2483 #2540
- [Merged] @grzi Reimplement the hide hierarchy system
- [Needs review / justification] @ezpuzz is working on a change to use system structs instead of build_xyz_system functions

**Comment by ezpuzz** [comment]
Remaining Tasks for Release:

- [ ] amethyst_animation #2549 
- [ ] amethyst_gltf
- [ ] amethyst_derive
- [ ] amethyst_test
- [ ] Examples
  - [ ] animation
  - [ ] auto_fov
  - [ ] custom_game_data
  - [ ] custom_ui
  - [ ] debug_lines
  - [ ] debug_lines_ortho
  - [ ] gltf
  - [ ] mouse_raycast
  - [ ] pong_tutorial_06
  - [ ] prefab_basic
  - [ ] prefab_custom
  - [ ] prefab_multi
  - [ ] renderable
  - [ ] renderable_custom
  - [ ] rendy
  - [ ] spotlights
  - [ ] sprite_animation
  - [ ] state_dispatcher
  - [ ] states_ui
  - [ ] ui
- [ ] docs and books #2557 
- [ ] website update @AlveLarsson

**Comment by grzi** [comment]
I'll give gltf port a shot for a couple of days. No guarantee though because it's new things for me (gltf && using new prefabs) 
Will keep track of my work on https://github.com/grzi/amethyst/issues/2

**Comment by grzi** [comment]
New tracking for GLTF port will be done on #2631

**Comment by grzi** [comment]
The GLTF port is ready at : https://github.com/amethyst/amethyst/pull/2631

**Comment by grzi** [comment]
Spotlights example port to legion : https://github.com/amethyst/amethyst/pull/2651

**Comment by B-Reif** [comment]
Hey, I'm just exploring amethyst for the first time, I'd be happy to take a stab at working on updating the book. Might need a little guidance.

**Comment by aakashhemadri** [comment]
Hey not sure if I should start using 0.15.3 with 0.16 around the corner, Do I wait for the 0.16 release?

**Comment by neogeek23** [comment]
In the same boat as @aakashhemadri but a few months later (not wanting to get started with 15.3 if 16.0 is close).  The checklist is looking nearly complete atm the only sub crate missing is amethyst_test.  Is it reasonable to expect this to release this year?

**Comment by CaesarXInsanium** [comment]
oh my goodness,its getting done yass
